### PR TITLE
Adding Prometheus scrape interval validation

### PR DESCRIPTION
**Pull request by avilevy18**
## Description

 Adding validator to Prometheus receivers config `scrape_interval` must be >= 5s
## Related issue

b/324071154

## Checklist:
- Unit tests
  - [ ] Unit tests do not apply.
  - [x] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [x] Integration tests do not apply.
  - [ ] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [x] This PR introduces no user visible changes.
  - [ ] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [x] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.
  - [ ] This PR bumps the version.

### CODE REVIEW

File: confgenerator/prometheus.go [hunk 1]
```diff
@@ -225,6 +228,9 @@ func validatePrometheus(promConfig promconfig.Config) (string, error) {
 	}
 
 	for _, sc := range promConfig.ScrapeConfigs {
+		if sc.ScrapeInterval < MinScrapeInterval {
+			return "scrape_interval", fmt.Errorf("error validation scrap_config for job %v: scrape_interval must be equal to or greater than %v", sc.JobName, MinScrapeInterval.String())
```
Comment (jefferbrecht) [inline confgenerator/prometheus.go:232]:
```suggestion
			return "scrape_interval", fmt.Errorf("error validating scrape_config for job %v: scrape_interval must be equal to or greater than %v", sc.JobName, MinScrapeInterval.String())
```

File: confgenerator/testdata/goldens/invalid-metrics-receiver_prometheus_scrape_interval/golden/linux-gpu/error [hunk 2]
```diff
@@ -0,0 +1 @@
+Key: 'PrometheusMetrics.config.config' Error:Field validation for 'config' failed on the 'error validation scrap_config for job prometheus: scrape_interval must be equal to or greater than 5s' tag
```
Comment (jefferbrecht) [inline confgenerator/testdata/goldens/invalid-metrics-receiver_prometheus_scrape_interval/golden/linux-gpu/error:1]: Please update the error files as well

File: confgenerator/prometheus.go [hunk 3]
```diff
@@ -229,7 +229,8 @@ func validatePrometheus(promConfig promconfig.Config) (string, error) {
 
 	for _, sc := range promConfig.ScrapeConfigs {
 		if sc.ScrapeInterval < MinScrapeInterval {
-			return "scrape_interval", fmt.Errorf("error validating scrape_config for job %v: scrape_interval must be equal to or greater than %v", sc.JobName, MinScrapeInterval.String())
+			sc.ScrapeInterval = MinScrapeInterval
+			log.Println("scrape_interval must be at least 5s; adjusting to minimum accepted value")
```
Comment (jefferbrecht) [inline confgenerator/prometheus.go:233]:
```suggestion
			log.Printf("scrape_interval must be at least %v; adjusting to minimum accepted value\n", MinScrapeInterval)
```

File: confgenerator/testdata/goldens/metrics-receiver_prometheus_scrape_interval/golden/windows/otel.yaml [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+    config:
+      global:
+        scrape_interval: 1m
+        scrape_timeout: 10s
+        evaluation_interval: 1m
+      scrape_configs:
+      - job_name: prometheus
+        honor_timestamps: true
+        scrape_interval: 5s
+        scrape_timeout: 1s
```
Comment (jefferbrecht) [inline confgenerator/testdata/goldens/metrics-receiver_prometheus_scrape_interval/golden/windows/otel.yaml:552]: The `prometheus1` job sets `scrape_timeout` to `6s`, presumably to match `scrape_interval`, but here it's still 1s. I can't find an explanation in the code though...

Comment (avilevy18) [reply confgenerator/testdata/goldens/metrics-receiver_prometheus_scrape_interval/golden/windows/otel.yaml:552]: Spoke offline.
Decision: if we clamp the `scrape_interval` we will also set `scrape_time` to `scrape_interval`

File: confgenerator/prometheus.go [hunk 5]
```diff
@@ -225,6 +228,11 @@ func validatePrometheus(promConfig promconfig.Config) (string, error) {
 	}
 
 	for _, sc := range promConfig.ScrapeConfigs {
+		if sc.ScrapeInterval < minScrapeInterval {
+			sc.ScrapeInterval = minScrapeInterval
+			sc.ScrapeTimeout = minScrapeInterval
+			log.Printf("scrape_interval must be at least %v; adjusting to minimum accepted value\n", minScrapeInterval)
```
Comment (jefferbrecht) [inline confgenerator/prometheus.go:234]: I think we should also log the fact that we're adjusting `scrape_timeout`, though admittedly I don't have a great idea in mind for the wording. Maybe we simply emit two separate logs:

```suggestion
			log.Printf("scrape_interval must be at least %v; adjusting to minimum accepted value\n", minScrapeInterval)
			log.Printf("adjusting scrape_timeout to %v\n", minScrapeInterval)
```

File: confgenerator/testdata/goldens/metrics-receiver_prometheus_scrape_interval/input.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        type: prometheus
+        config:
+          scrape_configs:
+          - job_name: prometheus
+            scrape_interval: 1s
+            static_configs:
+            - targets:
+              - localhost:1234
+          - job_name: prometheus1
+            scrape_interval: 6s
```
Comment (jefferbrecht) [inline confgenerator/testdata/goldens/metrics-receiver_prometheus_scrape_interval/input.yaml:27]: You probably meant for this to be above `minScrapeInterval`?

**Review by jefferbrecht — APPROVED** [review]

**Pull request merged**